test/AssetPool.test.js

```diff
@@ -975,4 +975,243 @@ describe("AssetPool", () => {
+            .connect(thirdParty)
+            .approveNewAssetPoolUpgrade(newAssetPool.address)
+        ).to.be.revertedWith("Ownable: caller is not the owner")
+      })
+    })
+  })
+
+  describe("upgradeToNewAssetPool", () => {
+    let newAssetPool
+    let newUnderwriterToken
+
+    beforeEach(async () => {
+      const NewUnderwriterToken = await ethers.getContractFactory(
+        "UnderwriterToken"
+      )
+      newUnderwriterToken = await NewUnderwriterToken.deploy(
+        "New Underwriter Token",
+        "newCOV"
+      )
+      await newUnderwriterToken.deployed()
+
+      const NewAssetPool = await ethers.getContractFactory("NewAssetPoolStub")
+      newAssetPool = await NewAssetPool.deploy(
+        collateralToken.address,
+        newUnderwriterToken.address
+      )
+      await newAssetPool.deployed()
+      await newUnderwriterToken.transferOwnership(newAssetPool.address)
+    })
+
+    context("when upgrading with zero underwriter tokens", () => {
+      it("should revert", async () => {
+        await expect(
+          assetPool.connect(underwriter1).upgradeToNewAssetPool(0)
+        ).to.be.revertedWith("Underwriter token amount must be greater than 0")
+      })
+    })
+
+    context("when upgrading with amount less than available", () => {
+      it("should revert", async () => {
+        const amount = to1e18(100)
+        await assetPool.connect(underwriter1).deposit(amount)
+        await underwriterToken
+          .connect(underwriter1)
+          .approve(assetPool.address, amount)
+
+        const amountToUpgrade = to1e18(101)
+        await expect(
+          assetPool.connect(underwriter1).upgradeToNewAssetPool(amountToUpgrade)
+        ).to.be.revertedWith(
+          "Underwriter token amount exceeds available balance"
+        )
+      })
+    })
+
+    context("when governance hasn't upgraded to a new pool yet", () => {
+      it("should revert", async () => {
+        const amount = to1e18(100)
+        await assetPool.connect(underwriter1).deposit(amount)
+        await underwriterToken
+          .connect(underwriter1)
+          .approve(assetPool.address, amount)
+
+        const amountToUpgrade = to1e18(99)
+        await expect(
+          assetPool.connect(underwriter1).upgradeToNewAssetPool(amountToUpgrade)
+        ).to.be.revertedWith("New asset pool must be assigned")
+      })
+    })
+
+    context("when governance upgraded to a new pool", () => {
+      const amountToDeposit = to1e18(100)
+      const amountToUpgrade = to1e18(40)
+
+      context("when no collateral tokens were claimed by the pool", () => {
+        let tx
+
+        beforeEach(async () => {
+          await assetPool
+            .connect(coveragePool)
+            .approveNewAssetPoolUpgrade(newAssetPool.address)
+
+          await assetPool.connect(underwriter1).deposit(amountToDeposit)
+          await underwriterToken
+            .connect(underwriter1)
+            .approve(assetPool.address, amountToDeposit)
```

Comment:
Shouldn't `amountToUpgrade` be approved instead?

Reply:
18fed51